Monika-After-Story/MonikaModDev
# An absolutely ordinary PR

<details>
  <summary>Dare to click?</summary>

A huge update for the *event*.

### Changes:
NOTE: I could've missed something because of the size, but in general all the changes are relevant, the only thing that's important is how the event works now.

 - Added `MASImageData`, a port of `ImageData` from the future renpy version, allows to load an image from binary data into memory. Useful!
 - Added 3 new vars to keep track of the progress of the event: `persistent._mas_islands_start_lvl`, `persistent._mas_islands_progress`, `persistent._mas_islands_unlocks`
 - Added `IslandsImageDefination` which is used to define new objects for the event, check the code to see how to add more stuff (although, we don't have/really want anything else to add)
 - Implemented randomized progression, the player unlocks new stuff as they spend time with Monika. Progression lvl updates once per day and on startup. Check `mas_island_event.__handleUnlocks` and `mas_island_event.advanceProgression` for more details
 - The users who have already seen the event before will get the highest currently available lvl (via an update script). The new users will have to unlock stuff and keep an eyes on the islands and how they change over time
 - Added new function which returns a composite image for the event `mas_island_event.getIslandsDisp`. The displayable itself isn't being cached between different calls, not worth it
 - Added new label `mas_islands`, in general it should be used to show the islands. It has in and out transitions, allows to show the image in static and interactive modes. Check `greeting_ourreality` and `mas_monika_islands` to see how it works
 - The old screens are commented out, added a new one `mas_islands`. You probably shouldn't show it directly and use the label instead
 - The old images have been removed. The new images are packed and encoded together, during decoding we load them in the memory, not on disk. This should speed things up massively, plus it obfuscates the event even more (but there's a little surprise for people who figure out how to decode it)
 - Added the sunset sprites by @kkrosie123 
 - Updated the code to support the sunset variant
 - Removed UI sprites which were used only in one screen during the event. They were inconsistent with the other UI elements. Now we use a similar style to the rest of the UI + reuse already existing assets (less files on disk too!)
 - Removed the window frame from the islands event, nobody liked it anyway (even less files on disk!)
 - Added `mas_island_event.shouldDecodeImages` as a generalized check for whether or not we should try to decode the images
 - New store namespace `mas_parallax` and the parallax sprites system 👀
 - - `ParallaxDecal` represents a single decal on a parallax sprite (multiple decals are composed together for optimization)
 - - `_ParallaxDecalContainer` is used internally, basically what the name says
 - - `ParallaxSprite` represents a parallax sprite. it's being built from an image and additional decals (if they are desired)
 - - `ParallaxBackground` is like RenPy's `Composite` but works 10 times faster, used when we need lots of parallax sprites on the screen
 - Removed `_mas_monika_islands_unlock` as unused
 - Updated `greeting_ourreality` and `mas_monika_islands` to use the new system + cleaned them up
 - Kinda deprecated `mas_cannot_decode_islands` and added `mas_decoded_islands` (double `not` is inconvenient to use in code). For now both are available
 - Added `EV_FLAG_DEF`, just for convenience so we don't use `0` in our code

### Fixed:
 - Fixed a crash in `mas_sprites._verify_mfwm`
 - Fixed a hotkeys/shields issue in `mas_monika_islands`, now you can't break the flow anymore

### Testing:
 - Verify general stability
 - Check the updated events (`greeting_ourreality` and `mas_monika_islands`), they should work as before, just provide better visuals
 - Try to see the islands at different levels (you start with a few islands, final lvl shows them all + all the decals)
 - Test update scripts, you should get the latest lvl (7 as of now) which should also unlock everything that's available for now
 - Verify nothing breaks if we can't decode the images (remove the file, corrupt it)
 - Try to interact with the islands, verify everything works as expected (smooth movement, proper click detection, no lag spikes)
 - Verify the event works as intended during sunset, check out new sprites

</details>
